src/Servers/Kestrel/Core/src/Internal/Http/HttpParser.cs

```diff
@@ -164,135 +176,144 @@ public bool ParseHeaders(TRequestHandler handler, ref SequenceReader<byte> reade
-
-                    KestrelBadHttpRequestException.Throw(RequestRejectionReason.InvalidRequestHeadersNoCRLF);
-                }
-
-                var length = 0;
-                // We only need to look for the end if we didn't read ahead; otherwise there isn't enough in
-                // in the span to contain a header.
-                if (readAhead == 0)
-                {
-                    length = span.IndexOfAny(ByteCR, ByteLF);
-                    // If not found length with be -1; casting to uint will turn it to uint.MaxValue
-                    // which will be larger than any possible span.Length. This also serves to eliminate
-                    // the bounds check for the next lookup of span[length]
-                    if ((uint)length < (uint)span.Length)
+                    else
                     {
-                        // Early memory read to hide latency
-                        var expectedCR = span[length];
-                        // Correctly has a CR, move to next
-                        length++;
-
-                        if (expectedCR != ByteCR)
+                        // What's after the CR?
+                        if (!reader.TryPeek(out _))
                         {
-                            // Sequence needs to be CRLF not LF first.
-                            RejectRequestHeader(span[..length]);
+                            // No more chars after CR? Don't consume an incomplete header
+                            reader.Rewind(crIndex + 1);
+                            return false;
                         }
-
-                        if ((uint)length < (uint)span.Length)
+                        else if (crIndex == 0)
                         {
-                            // Early memory read to hide latency
-                            var expectedLF = span[length];
-                            // Correctly has a LF, move to next
-                            length++;
-
-                            if (expectedLF != ByteLF ||
-                                length < 5 ||
-                                // Exclude the CRLF from the headerLine and parse the header name:value pair
-                                !TryTakeSingleHeader(handler, span[..(length - 2)]))
-                            {
-                                // Sequence needs to be CRLF and not contain an inner CR not part of terminator.
-                                // Less than min possible headerSpan of 5 bytes a:b\r\n
-                                // Not parsable as a valid name:value header pair.
-                                RejectRequestHeader(span[..length]);
-                            }
-
-                            // Read the header successfully, skip the reader forward past the headerSpan.
-                            span = span.Slice(length);
-                            reader.Advance(length);
+                            // CR by itself
```

Comment:
Wouldn't a CR by itself be fine if it was followed by an LF? It would mark the end of the headers, right?

Reply:
There's a fastpath above that catches CR+LF at the start so this is CR+notLF.

Reply:
Probably worth mentioning that in the comment though :smile:

Reply:
The LF could still arrive in a subsequent receive call and that should be valid.

Reply:
> in a subsequent receive call

In a separate span that is? Which is what is tested above, in the same span or in different spans

```c#
  if ((uint)span.Length > (uint)(crIndex + 1) && span[crIndex + 1] == ByteLF)
  {
      // CR/LF in the same span (common case)
      span = span.Slice(0, crIndex);
      foundCrlf = true;
  }
  else if (reader.TryPeek(out byte lfMaybe) && lfMaybe == ByteLF)
  {
      // CR/LF but split between spans
      span = span.Slice(0, span.Length - 1);
      foundCrlf = true;
  }
```

Reply:
It's not that it's a separate. The final LF may have simply not have arrived yet. Maybe it hasn't even been sent. We should wait to see what the next byte will be.

Reply:
Is there a test that simulates that already? Functional tests maybe?

Reply:
As I wrote "delay" I think I remember which one does that

Reply:
There is a `TryPeek` right before, and if it fails then we return `false`, isn't that the case you are describing?

Reply:
> There is a `TryPeek` right before, and if it fails then we return `false`, isn't that the case you are describing?

I don't know how I missed that. It's well commented too.

Reply:
We can improve this comment though because the CR is not "by itself" in this case. This is the case where we know there is another byte that is not LF.

```suggestion
                            // CR followed by something other than LF
```